core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1984,6 +1984,36 @@ object Stream {
       self.reduce(S.combine(_, _))
 
     /**
+      * Repartitions the input with the function `p`. On each step `p` is applied
+      * to the input and all elements but the last of the resulting sequence
+      * are emitted. The last element is then appended to the next input using the
+      * Semigroup `S`.
+      *
+      * @example {{{
+      * scala> import cats.implicits._
+      * scala> Stream("Hel", "l", "o Wor", "ld").repartition(_.split(" ").toIndexedSeq).toList
+      * res0: List[String] = List(Hello, World)
+      * }}}
+      */
+    def repartition(p: O => IndexedSeq[O])(implicit S: Semigroup[O]): Stream[F,O] = {
+      def go(carry: Option[O], s: Stream[F,O]): Pull[F,O,Unit] = {
+        s.pull.uncons1.flatMap {
```

Comment:
It would be much more efficient to use `s.pull.uncons` here. It will complicate the body somewhat but likely worth the tradeoff.